File: scratch/record-scalarization-integration-20260916/final/src/Test/RecordsFFI.cc
#include "purescript.h"
#include <memory>

namespace {
    struct Inner { int e, f; };
    struct Middle { int c; std::shared_ptr<const Inner> d; };
    struct Record { int a; std::shared_ptr<const Middle> b; };
    using RecordPtr = std::shared_ptr<const Record>;

    RecordPtr initial() {
        const auto d = std::make_shared<Inner>(Inner{0, 0});
        const auto b = std::make_shared<Middle>(Middle{0, d});
        return std::make_shared<Record>(Record{0, b});
    }

    RecordPtr updateRec(int n, RecordPtr record) {
        while (n > 0) {
            const auto d = std::make_shared<Inner>(Inner{
                record->b->d->e + 3, record->b->d->f + n % 5});
            const auto b = std::make_shared<Middle>(Middle{record->b->c + 2, d});
            record = std::make_shared<Record>(Record{record->a + 1, b});
            --n;
        }
        return record;
    }
}

FOREIGN_BEGIN(Test_RecordsFFI)
exports["runRecordsFFI"] = [](const boxed& in) -> boxed {
    const auto result = updateRec(unbox<int>(in), initial());
    const int value = result->b->d->f;
    return value;
};
FOREIGN_END

File: scratch/compact-guards-integration-20260916/compact/src/Test/RecordsFFI.fs
module Test.RecordsFFI

type Inner = { e: int; f: int }
type Middle = { c: int; d: Inner }
type DeepRecord = { a: int; b: Middle }
let rec update n r =
    if n = 0 then r
    else update (n - 1)
            { a = r.a + 1
              b = { c = r.b.c + 2; d = { e = r.b.d.e + 3; f = r.b.d.f + n % 5 } } }
let runRecordsFFI (n: obj) =
    let initial = { a = 0; b = { c = 0; d = { e = 0; f = 0 } } }
    (update (unbox<int> n) initial).b.d.f :> obj
